warehouse/oidc/models/github.py

```diff
@@ -235,6 +241,25 @@ def publisher_url(self, claims=None):
             return f"{base}/commit/{sha}"
         return base
 
+    def publisher_verification_policy(self, claims):
+        sha = claims.get("sha") if claims else None
+        ref = claims.get("ref") if claims else None
+        if not (ref or sha):
+            raise InvalidPublisherError("The ref and sha claims are empty")
+
+        expected_build_configs = [
+            OIDCBuildConfigURI(f"https://github.com/{self.job_workflow_ref}@{claim}")
+            for claim in [ref, sha]
+            if claim is not None
+        ]
```

Comment:
Nitpick: I might be mistaken, but in practice this should only be `ref`, not `sha`. With that being said, we can probably use `OIDCSourceRepositoryDigest` to enforce that the `sha` also matches 🙂

Reply:
The trusted publisher claim verification attempts both though: https://github.com/pypi/warehouse/blob/784f446e1ebedfaceda2404efcf76ea29f9d80e2/warehouse/oidc/models/github.py#L56-L62

Reply:
Thinking about this further, the current claims (plus the Digest one you suggested above) are:
1. OIDCBuildConfigURI (e.g: https://github.com/org/repo/.github/workflows/workflow.yml@$REF/$SHA)
2. OIDCSourceRepositoryURI (e.g: https://github.com/org/repo/)
3. OIDCSourceRepositoryDigest (the commit SHA)

I think we could remove the repo URI (2), since it's included in the build config URI (1). WDYT?

Reply:
Makes sense to me!

Reply:
done!

Reply:
Just tying things together: https://github.com/pypi/warehouse/pull/14335 is where checking both the `ref` and the `sha` was introduced.

Reply:
Here's a PR that adds a comment explaining why we check for both: https://github.com/pypi/warehouse/pull/15967